Add section on input limits for psk, pskID, info, exporter_context (#117)
cfrg/draft-irtf-cfrg-hpke: draft-irtf-cfrg-hpke.md

That is, if a non-default value is provided for one of them, then
the other MUST be set to a non-default value.

The `psk`, `pskID`, and `info` fields have maximum lengths that depend
on the KDF itself, on the definition of `LabeledExtract`, and on the
constant labels used together with them. See {{kdf-input-length}} for

> exact indications.

Reviewer: ```suggestion
precise limits on these lengths.
```

a secret derived from the internal exporter secret using the corresponding KDF Expand
function.

The `exporter_context` field has a maximum length that depends on the KDF
itself, on the definition of `LabeledExpand`, and on the constant labels

> used together with them. See {{kdf-input-length}} for exact indications.

Reviewer: ```suggestion
used together with them. See {{kdf-input-length}} for exact limits on this length.
```

| 0x0002 | HKDF-SHA384 | 48  | {{?RFC5869}} |
| 0x0003 | HKDF-SHA512 | 64  | {{?RFC5869}} |

### Input Length Restrictions {#kdf-input-length}

This document defines `LabeledExtract` and `LabeledExpand` based on the
KDFs listed above. These functions add prefixes to their respective
inputs `IKM` and `info` before calling the KDF's Extract and Expand
functions. This leads to a reduction of the maximum input length that

> is available for the inputs `psk`, `pskID`, `info`, `exporter_context`.
> The following table lists the maximum allowed lengths of these fields

Reviewer: ```suggestion
is available for the inputs `psk`, `pskID`, `info`, `exporter_context`,
the variable-length parameters provided by HPKE applications.
The following table lists the maximum allowed lengths of these fields
```

for the KDFs defined in this document, as inclusive bounds in bytes:

| Input            | HKDF-SHA256  | HKDF-SHA384   | HKDF-SHA512   |
|:-----------------|:-------------|:--------------|:--------------|
| psk              | 2^{61} - 81  | 2^{125} - 145 | 2^{125} - 145 |
| pskID            | 2^{61} - 83  | 2^{125} - 147 | 2^{125} - 147 |
| info             | 2^{61} - 82  | 2^{125} - 146 | 2^{125} - 146 |

> | exporter_context | 2^{61} - 111 | 2^{125} - 191 | 2^{125} - 207 |

Reviewer 2: While I guess it's worth knowing that these limits exist, they seem ... unlikely to ever be hit in reality.

Author: That's right. I added a couple of words to put it into perspective.

The values for `psk`, `pskID`, and `info` which are inputs to

> `LabeledExtract` were computed with the following equation:

Reviewer: ```suggestion
`LabeledExtract` were computed with the following expression:
```

> ~~~
> max_size_hash_input - Nb - size_label_rfcXXXX - size_label
> ~~~
> 
> The value for `exporter_context` which is an input to `LabeledExpand`
> was computed with the following equation:

Reviewer: ```suggestion
~~~
max_size_hash_input - Nb - size_label_rfcXXXX - size_input_label
~~~

The value for `exporter_context` which is an input to `LabeledExpand`
was computed with the following expression:
```

~~~

> max_size_hash_input - Nb - Nh - size_label_rfcXXXX - size_label - 2 - 1

Reviewer: ```suggestion
max_size_hash_input - Nb - Nh - size_label_rfcXXXX - size_input_label - 2 - 1
```

~~~

In these equations, `max_size_hash_input` is the maximum input length
of the underlying hash function in bytes, `Nb` is the block size of the
underlying hash function in bytes, `size_label_rfcXXXX` is the size

> of "RFCXXXX " in bytes and equals 8, and `size_label` is the size of

Reviewer: ```suggestion
of "RFCXXXX " in bytes and equals 8, and `size_input_label` is the size of
```
